compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1880,21 +1881,103 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
-    fn no_such_field_err<T: Display>(
+    fn no_such_field_err(
         &self,
+        base_span: Span,
         span: Span,
-        field: T,
-        expr_t: &ty::TyS<'_>,
+        field: Ident,
+        expr_t: &'tcx ty::TyS<'tcx>,
     ) -> DiagnosticBuilder<'_> {
-        type_error_struct!(
+        debug!("no_such_field_err(span: {:?}, field: {:?}, expr_t: {:?})", span, field, expr_t);
+
+        let mut err = type_error_struct!(
             self.tcx().sess,
             span,
             expr_t,
             E0609,
             "no field `{}` on type `{}`",
             field,
             expr_t
-        )
+        );
+
+        // try to add a suggestion in case the field is a nested field of a field of the Adt
+        if let Some((fields, substs)) = self.get_field_candidates(span, &expr_t) {
+            for candidate_field in fields.iter() {
+                if let Some(field_path) =
+                    self.check_for_nested_field(span, field, candidate_field, substs, vec![])
+                {
+                    let base_sp = base_span.with_hi(field.span.lo());
+                    let mut suggestion_str =
+                        self.tcx.sess.source_map().span_to_snippet(base_sp).unwrap();
+                    let field_path_str = field_path
+                        .iter()
+                        .map(|id| id.name.to_ident_string())
+                        .collect::<Vec<String>>()
+                        .join(".");
+                    debug!("field_path_str: {:?}", field_path_str);
+                    suggestion_str.push_str(&field_path_str);
+
+                    err.span_suggestion(
+                        span,
+                        "try to use the whole path to the field",
+                        suggestion_str,
+                        Applicability::MaybeIncorrect,
+                    );
```

Comment:
This is what I would do instead:

```suggestion
                    err.span_suggestion_verbose(
                        ident.span.shrink_to_lo(),
                        "one of the expressions fields has a field of the same name",
                        format!(".{}", field_path.iter().map(|i| i.to_string()).collect::<Vec<_>>().join(".")),
                        Applicability::MaybeIncorrect,
                    );
```